### ISSUE TITLE
No documentation on how a server must response to a wrong CORS-headers-request.

### ISSUE THREAD

**Issue by drafu** [whatwg/fetch#1102]
The specification states that a client must send `Access-Control-Request-Method` and `Access-Control-Request-Headers` headers to the server to "indicate which method/headers a future CORS request to the same resource might use".

But there is nowhere to be found explicitly, why a server needs them and how it should react if these client-sent headers are not included in the response.

What I've found:
- A client must send `Access-Control-Request-Method` and `Access-Control-Request-Headers`.
- A server must respond with available `Access-Control-Allow-Methods` and `Access-Control-Allow-Headers`
- A server must respond with 2xx status if it allows CORS requests.
- A server must respond with 403 status if it doesn't allow any CORS requests.

But nothing states about what server should respond with if it allows CORS requests, but requested method or headers are not available.

Example request/response:

REQUEST:
Access-Control-Request-Method: BADMETHOD
Access-Control-Request-Headers: Bad-Header

RESPONSE: 200 OK (204 NO CONTENT)
Access-Control-Allow-Methods: OPTIONS,HEAD,GET,PUT,POST,DELETE
Access-Control-Allow-Headers: Origin,X-Requested-With,Content-Type,Accept

This seems illogical because these headers are no use for the server: the server has a static response no matter what requested headers were.

Thank you in advance.

**Comment by annevk** [comment]
The server can use them and make decisions based on the information provided by the client, but isn't required to. We don't even require a 403, just encourage it.

I'm happy to include some more advice, but if servers just want to echo a static reply to the OPTIONS request that ought to be okay and might well be more efficient than having to validate each request. Would stipulating that validating the request and potentially replying with a 403 is up to servers help?

**Comment by rossabaker** [comment]
I have the same question and did a [survey of various server implementations](https://github.com/http4s/http4s/issues/5322#issuecomment-933547345).  The prior [W3C spec](https://www.w3.org/TR/2020/SPSD-cors-20200602/#resource-preflight-requests) was more prescriptive on the server side.  I'm curious whether the Fetch API Spec is looser because those rules were deemed too prescriptive, or because that spec is just more focused on the client.

**Comment by annevk** [comment]
It seemed too prescriptive. Fetch has https://fetch.spec.whatwg.org/#http-cors-protocol instead. FWIW, I wrote both, I wouldn't read too much into the respective style sheets here.

**Comment by rossabaker** [comment]
Thanks, @annevk.  So there shouldn't be any security concern with an implementation returning CORS response headers even when one of the `Access-Control-Request-Headers` values is not allowed, or when the `Access-Control-Request-Method` method is not allowed.  It's just a hint to the server that's most helpful in the case of unbounded allowed values.

**Comment by annevk** [comment]
Yeah, your preflight response can be static and this might well make sense if you have caching intermediaries.
